### PR TITLE
BDT-101094: Create Azure key vault app from template

### PR DESCRIPTION
**Pull request by mrfreeman**

### CODE REVIEW

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/FetchCredentialsExecutionService.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            log.info(SUCCESSFULLY_FETCHED_SECRET);
+
+            return Triple.of(true, SUCCESSFULLY_FETCHED_SECRET, Map.of(secretKey, secretValue));
+
+        } catch (RuntimeException | JsonProcessingException e) {
+            bigIDProxy.updateActionStatusToBigID(
+                    executionContext,
+                    initializeResponse(executionContext, StatusEnum.ERROR, 0d, FAILED_TO_FETCH_SECRET)
+            );
+            log.error("Failed to fetch secret: {}", e.getMessage(), e);
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/FetchCredentialsExecutionService.java:74]: can use FAILED_TO_FETCH_SECRET in log as well

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/AzureKeyVaultTokenService.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        Map<String, String> globalParamsMap = ParamsMapUtils.getGlobalParamsMap(executionContext);
+
+        ClientSecretCredential credential = new ClientSecretCredentialBuilder()
+                .clientId(globalParamsMap.get(GlobalParams.CLIENT_ID.getValue()))
+                .clientSecret(GlobalParams.CLIENT_SECRET.getValue())
+                .tenantId(GlobalParams.TENANT_ID.getValue())
+                .build();
+
+        // Define the scope for Azure Key Vault
+        String scope = "https://vault.azure.net/.default";
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/AzureKeyVaultTokenService.java:29]: should scope be another part of the query param?

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/controller/LogoController.java [hunk 3]
```diff
@@ -1,14 +1,15 @@
-package com.basicapp.basicdemoapp.controllers;
+package com.bigid.azurekeyvaultapp.controller;
 
 import com.bigid.appinfrastructure.controllers.AbstractLogoController;
 import org.springframework.stereotype.Controller;
 
 @Controller
 public class LogoController extends AbstractLogoController {
-    public String getSideBarIconPath(){
+    public String getSideBarIconPath() {
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/controller/LogoController.java:8]: lets find an icon to use- might have one in core already

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/FetchCredentialsExecutionService.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import org.apache.commons.lang3.tuple.Triple;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Mono;
+
+import java.util.Map;
+
+@Service
+@Slf4j
+public class FetchCredentialsExecutionService extends AbstractExecutionService implements ExecutionService {
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/FetchCredentialsExecutionService.java:28]: for each of the actions lets add a validation phase for the params as the user can input invalid names/miss required fields and return useful errors for those as well

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/util/ParamsMapUtils.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package com.bigid.azurekeyvaultapp.util;
+
+import com.bigid.appinfrastructure.dto.ActionParamDetails;
+import com.bigid.appinfrastructure.dto.ExecutionContext;
+import com.bigid.appinfrastructure.dto.ParamDetails;
+
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ParamsMapUtils {
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/util/ParamsMapUtils.java:10]: Nice!

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/FetchCredentialsExecutionService.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    private final KeyVaultTokenService keyVaultTokenService;
+
+    @Autowired
+    public FetchCredentialsExecutionService(BigIDProxy bigIDProxy, KeyVaultTokenService keyVaultTokenService) {
+        super(bigIDProxy);
+        this.keyVaultTokenService = keyVaultTokenService;
+    }
+
+    @Override
+    public Triple<Boolean, String, Map<String, String>> performAction(ExecutionContext executionContext) {
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/FetchCredentialsExecutionService.java:45]: Might be nice to create a AzureVaultResponseClass with overrides for getLeft, getMiddle, getRight with the appropriate names for the fields to help with readability in the controller

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/FetchCredentialsExecutionService.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            String secretKey = rootNode.get(SECRET_KEY).asText();
+
+            String secretValue = secretClient.getSecret(secretKey).getValue();
+            bigIDProxy.updateActionStatusToBigID(
+                    executionContext,
+                    initializeResponse(executionContext, StatusEnum.COMPLETED, 1d, SUCCESSFULLY_FETCHED_SECRET)
+            );
+            log.info(SUCCESSFULLY_FETCHED_SECRET);
+
+            return Triple.of(true, SUCCESSFULLY_FETCHED_SECRET, Map.of(secretKey, secretValue));
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/FetchCredentialsExecutionService.java:67]: value returned should have shape: { credentialFields: {secretValue} } with credentialFields key word (assuming secretValue is the key:value set of the returned secret

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/FetchCredentialsExecutionService.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+                    .credential(tokenCredential)
+                    .buildClient();
+
+            // Fetch the secret from Azure Key Vault
+            ObjectMapper objectMapper = new ObjectMapper();
+            JsonNode rootNode = objectMapper.readTree(actionParamsMap.get(ActionParams.CREDENTIAL_PROVIDER_CUSTOM_QUERY.getValue()).toString());
+            String secretKey = rootNode.get(SECRET_KEY).asText();
+
+            String secretValue = secretClient.getSecret(secretKey).getValue();
+            bigIDProxy.updateActionStatusToBigID(
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/FetchCredentialsExecutionService.java:61]: not sure we need this update as the response in the controller will return success wdyt?

File: services/bigid-azure-key-vault-integration-app/src/main/resources/Manifest [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+      "param_priority": "primary",
+      "is_mandatory": true
+    },
+    {
+      "param_name": "authentication_method",
+      "param_friendly_name": "Authentication Method",
+      "param_type": "String",
+      "is_cleartext": true,
+      "param_description": "Authentication method determines how to authenticate with Hashicorp vault",
+      "default_value": "App Role",
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/main/resources/Manifest:37]: Client Credentials

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/AzureKeyVaultTokenService.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.Map;
+import java.util.Objects;
+
+@Service
+public class AzureKeyVaultTokenService implements KeyVaultTokenService {
+    @Override
+    public AccessToken fetchAccessToken(ExecutionContext executionContext) {
+        Map<String, String> globalParamsMap = ParamsMapUtils.getGlobalParamsMap(executionContext);
+
+        ClientSecretCredential credential = new ClientSecretCredentialBuilder()
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/AzureKeyVaultTokenService.java:22]: lets create an option for username/password auth as well

**Review by jacob-yeger — CHANGES_REQUESTED** [review]

File: services/bigid-azure-key-vault-integration-app/src/test/java/com/bigid/azurekeyvaultapp/AzureKeyVaultApplicationTest.java [hunk 11]
```diff
@@ -1,13 +1,13 @@
...
 
 @SpringBootTest
-class BasicDemoAppApplicationTests {
+class AzureKeyVaultApplicationTest {
 
-	@Test
-	void contextLoads() {
-	}
+    @Test
+    void contextLoads() {
```
Comment (jacob-yeger) [inline services/bigid-azure-key-vault-integration-app/src/test/java/com/bigid/azurekeyvaultapp/AzureKeyVaultApplicationTest.java:10]: lets add unit tests

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/AzureKeyVaultTokenService.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+                : new UsernamePasswordCredentialBuilder()
+                .clientId(globalParamsMap.get(GlobalParams.CLIENT_ID.getValue()))
+                .tenantId(globalParamsMap.get(GlobalParams.TENANT_ID.getValue()))
+                .username(globalParamsMap.get(GlobalParams.USERNAME.getValue()))
+                .password(globalParamsMap.get(GlobalParams.PASSWORD.getValue()))
+                .build();
+
+        return Objects.requireNonNull(credential
+                .getToken(new TokenRequestContext()
+                        .addScopes(globalParamsMap.get(GlobalParams.SCOPE.getValue())))
```
[the remark this answers is not in the document]
Comment (mrfreeman) [reply services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/AzureKeyVaultTokenService.java:38]: > is scope a global field or should it be per secret?

In Azure Key Vault, the scope is global and applies to the entire Key Vault instance, not individual secrets. When you specify a scope such as https://vault.azure.net/.default, it refers to the Azure Key Vault resource as a whole, not to specific secrets within it.

File: services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/AzureKeyVaultTokenService.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        Map<String, String> globalParamsMap = ParamsMapUtils.getGlobalParamsMap(executionContext);
+
+        TokenCredential credential = globalParamsMap.get(GlobalParams.AUTHENTICATION_METHOD.getValue()).equals(GlobalParams.CLIENT_SECRET.getValue())
+                ? new ClientSecretCredentialBuilder()
+                .clientId(globalParamsMap.get(GlobalParams.CLIENT_ID.getValue()))
+                .clientSecret(globalParamsMap.get(GlobalParams.CLIENT_SECRET.getValue()))
+                .tenantId(globalParamsMap.get(GlobalParams.TENANT_ID.getValue()))
+                .build()
+                : new UsernamePasswordCredentialBuilder()
+                .clientId(globalParamsMap.get(GlobalParams.CLIENT_ID.getValue()))
```
[the remark this answers is not in the document]
Comment (mrfreeman) [reply services/bigid-azure-key-vault-integration-app/src/main/java/com/bigid/azurekeyvaultapp/service/impl/AzureKeyVaultTokenService.java:30]: > need clientId for username/pass as well

Yes, we still need the Client ID

File: services/bigid-azure-key-vault-integration-app/src/test/java/com/bigid/azurekeyvaultapp/AzureKeyVaultApplicationTest.java (hunk 11 above)
Comment (mrfreeman) [reply services/bigid-azure-key-vault-integration-app/src/test/java/com/bigid/azurekeyvaultapp/AzureKeyVaultApplicationTest.java:10]: > lets add unit tests

will add right after the doc and manual testing

**Review by jacob-yeger — APPROVED** [review]

**Pull request merged**